File: Lesson11_ClassRelationships/01_RelationshipsExample/RelationshipsExample.java
/*
Copyright Ann Barcomb and Emily Marasco, 2023
Licensed under GPL v3
See LICENSE.txt for more information.
*/
package edu.ucalgary.oop;

public class RelationshipsExample{
 
    public static void main(String[] args) {
        
    //Build a new house.
    House newBuild = new House("Pat Neil", "123 Main St. NW", "cedar", "2010", 10.0, 6, 3, 4.5, "inset");

    //Purchase a new door.
    Door newDoor = new Door("metal","green");
    
    //Purchase a new bathtub.
    Bathtub newTub = new Bathtub("clawfoot","blue");

    //Make renovations.
    newBuild.buildDeck("wood", 37.5, "2021");
    newBuild.renovateBathroom("farmhouse", newTub);
    newBuild.updateRoof("asphalt", "2021");
    newBuild.replaceDoor("front", newDoor);
    
    //Find tenant(s).
    Tenant renter1 = new Tenant("Jordan Upton","engineer","587-289-1111");
    Tenant renter2 = new Tenant("Jamie Upton","teacher","403-787-2222",renter1);
    Tenant[] renterInfo = new Tenant[]{renter1, renter2};
    
    //Sign rental agreement.
    RentalAgreement contract = new RentalAgreement("January 2021", "June 2021", 2500.00, renterInfo, newBuild);
    contract.rentalSummary();
    RentalAgreementHardcopy legalCopy = new RentalAgreementHardcopy();
    legalCopy.printAgreement(contract);
    }    
}


class House {
    
    private String owner;
    private final String ADDRESS;
    private Roof houseRoof;
    private Kitchen houseKitchen;
    private Bathroom houseBathroom;
    private Door frontDoor;
    private Door backDoor;
    private Deck houseDeck;
    
    
	// Create a new house with default doors and no deck
    public House(String owner, String address, String roofMaterial, String roofInstalled, double kitchenArea, int kitchenSeats, int numAppliances, double bathroomArea, String sinkType){
        this.owner = owner;
        this.ADDRESS = address;
        this.houseRoof = new Roof(roofMaterial, roofInstalled);
        this.houseKitchen = new Kitchen(kitchenArea, kitchenSeats, numAppliances);
        this.houseBathroom = new Bathroom(bathroomArea, sinkType);
		this.frontDoor = new Door();
		this.backDoor = new Door();
    }

	// Create new house with custom doors and no deck
    public House(String owner, String address, String roofMaterial, String roofInstalled, double kitchenArea, int kitchenSeats, int numAppliances, double bathroomArea, String sinkType, Door frontDoor, Door backDoor){
        this.owner = owner;
        this.ADDRESS = address;
        this.houseRoof = new Roof(roofMaterial, roofInstalled);
        this.houseKitchen = new Kitchen(kitchenArea, kitchenSeats, numAppliances);
        this.houseBathroom = new Bathroom(bathroomArea, sinkType);
        this.frontDoor = frontDoor;
        this.backDoor = backDoor;
    }
    
	// Update the roof material and year
    public void updateRoof(String roofMaterial, String roofInstalled){
        this.houseRoof = new Roof(roofMaterial, roofInstalled);
    }
    
	// Retrieve current roof installation year - this is set to a new value whenever the roof is updated through the setter in Roof
    public String getInstallationYear(){
        return this.houseRoof.getINSTALLATIONYEAR();
    }
    
	// Update the bathroom tub and sink
    public void renovateBathroom(String sinkType, Bathtub newTub){
        this.houseBathroom.addTub(newTub);
        this.houseBathroom.setSink(sinkType);
    }
 
	// Update the deck material and year
    public void buildDeck(String material, double area, String installationYear){
        this.houseDeck = new Deck(material, area, installationYear);
    }

	// Change the front or back door
    public void replaceDoor(String whichDoor, Door newDoor){
        if(whichDoor.equalsIgnoreCase("front")){
            this.frontDoor = newDoor;
        }else if(whichDoor.equalsIgnoreCase("back")){
            this.backDoor = newDoor;
        }
    }
    
	// Retrieve current owner
    public String getOwner(){
        return this.owner;
    }

	// Update current owner
    public void setOwner(String owner){
        this.owner = owner;
    }

	// Return the address - address cannot be changed, so no setter is needed
    public String getADDRESS(){
        return this.ADDRESS;
    }


	// Convention is to have getters for everything. Setters might not be needed depending on the functionality- in this case, most of the reno methods are the setters, but they are named for their behaviour.
    public Roof getRoof(){
        return this.houseRoof;
    }

    public Kitchen getKitchen(){
        return this.houseKitchen;
    }
	
    public Bathroom getBathroom(){
        return this.houseBathroom;
    }
    
	public Door getFrontDoor(){
        return this.frontDoor;
    }
    
	public Door getBackDoor(){
        return this.backDoor;    
	}
    
	public Deck getDeck(){
        return this.houseDeck;
    }
        
}


// Composition: A roof cannot exist without a house
class Roof{
    
    private final String MATERIAL;
    private final String INSTALLATIONYEAR;
    
    public Roof(String material, String installationYear){
        this.MATERIAL = material;
        this.INSTALLATIONYEAR = installationYear;
    }

    public String getINSTALLATIONYEAR(){
        return this.INSTALLATIONYEAR;
    }
    
    public String getMATERIAL(){
        return this.MATERIAL;
    }
}

// Composition: A deck cannot exist without a house
class Deck{
    
    private final String MATERIAL;
    private double area;
    private String installationYear;
    
    public Deck(String material, double area, String installationYear){
        this.MATERIAL = material;
        this.area = area;
        this.installationYear = installationYear;
        System.out.println("Enjoy your new deck!");
    }    
    
    public String getMATERIAL(){
        return this.MATERIAL;
    }
    
    public double getArea(){
        return this.area;
    }

    public void setArea(double area){
        this.area = area;
    }
    
    public String getInstallYear(){
        return this.installationYear;
    }

    public void setInstallYear(String installYear){
        this.installationYear = installYear;
    }

}

// Aggregation: Doors can be purchased and moved between houses- therefore, they are part of the house, but they can exist even if the house does not.
class Door{
    private final String MATERIAL;
    private final String COLOUR;
    
    public Door(){  
        this.MATERIAL = "wood";
        this.COLOUR = "white";
    }
    
    public Door(String material, String colour){
        this.MATERIAL = material;
        this.COLOUR = colour;
    }
    
    public String getMATERIAL(){
        return this.MATERIAL;
    }
    
    public String getCOLOUR(){
        return this.COLOUR;
    }
}

// Composition: A kitchen cannot exist without the house.
class Kitchen{

    private final double AREA;
    private int numSeats;
    private int numAppliances;
    
    public Kitchen(double area, int numSeats, int numAppliances){
        this.AREA = area;
        this.numSeats = numSeats;
        this.numAppliances = numAppliances;
    }
    
    public double getAREA(){
        return this.AREA;
    }
    
    public int getSeats(){
        return this.numSeats;
    }
    
	public void setSeats(int numSeats){
		this.numSeats = numSeats;
	}
	
    public int getNumAppliances(){
        return this.numAppliances;
    }

	public void setNumAppliances(int numAppliances){
		this.numAppliances = numAppliances;
	}

}

// Composition: A bathroom cannot exist without the house.
class Bathroom{

    private final double AREA;
    private String sinkType;
    private Bathtub tub;
    
    public Bathroom(double area, String sinkType){
        this.AREA = area;
        this.sinkType = sinkType;
    }

    public Bathroom(double area, String sinkType, Bathtub tub){
        this.AREA = area;
        this.sinkType = sinkType;
        this.tub = tub;
    }
    
    public void addTub(Bathtub newTub){
        this.tub = newTub;
        System.out.println("Enjoy your new tub!");        
    }
    
    public void setSink(String sinkType){
        this.sinkType = sinkType;
        System.out.println("New sink looks great!");        
    }
    
    public double getAREA(){
        return this.AREA;
    }
    
    public String getSinkType(){
        return this.sinkType;
    }
    
    public Bathtub getTub(){
        return this.tub;
    }
}


// Aggregation: Tubs can be purchased and moved between houses- therefore, they are part of the house, but they can exist even if the house does not.
class Bathtub{

    private final String TYPE;
    private final String COLOUR;
    
    public Bathtub(String type, String colour){
        this.TYPE = type;
        this.COLOUR = colour;
    }
    
    public String getTYPE(){
        return this.TYPE;
    }
    
    public String getCOLOUR(){
        return this.COLOUR;
    }
}

// Association: Tenants and houses have a relationship, but both can exist independently and are not part of each other.
class Tenant{
    
    private String name;
    private String occupation;
    private String phoneNumber;
    private Tenant houseContact;	// Reflexive Association: One tenant is responsible for the other tenants- this means some of the tenants will contain another tenant object. If the tenant IS the contact, that variable will remain null.
    
    public Tenant(String name, String occupation, String phoneNumber){
        this.name = name;
        this.occupation = occupation;
        this.phoneNumber = phoneNumber;        
    }

    public Tenant(String name, String occupation, String phoneNumber, Tenant houseContact){
        this.name = name;
        this.occupation = occupation;
        this.phoneNumber = phoneNumber;
        this.houseContact = houseContact;
    }
    
    public String getName(){
        return this.name;
    }

	public void setName(String name){
		this.name = name;
	}
    
    public String getOccupation(){
        return this.occupation;
    }
    
	public void setOccupation(String occupation){
		this.occupation = occupation;
	}
	
    public String getPhoneNumber(){
        return this.phoneNumber;
    }
    
	public void setPhoneNumber(String number){
		this.phoneNumber = number;
	}
	
    public Tenant getHouseContact(){
        return this.houseContact;
    }
	
	public void setHouseContact(Tenant contact){
		this.houseContact = contact;
	}
}


// Association class: Tenants and houses are linked using a rental agreement.
class RentalAgreement{
    
    private final String STARTDATE;
    private final String ENDDATE;
    private final double MONTHLYCOST;
    private final Tenant[] RENTERS;
    private final House RESIDENCE;
    
    public RentalAgreement(String startDate, String endDate, double monthlyCost, Tenant[] renters, House residence){
        this.STARTDATE = startDate;
        this.ENDDATE = endDate;
        this.MONTHLYCOST = monthlyCost;
        this.RENTERS = renters;
        this.RESIDENCE = residence;
    }
    
    public void rentalSummary(){
        System.out.println("The details of the agreement are as follows: ");
        System.out.println("Agreement is between: " + getBothParties());
        System.out.println("Begins: " + STARTDATE + " Ends: " + ENDDATE);
        System.out.println("Monthly cost: " + String.format("$%.2f", MONTHLYCOST));
        System.out.println("Roof was last updated: " + RESIDENCE.getInstallationYear());
        System.out.println();
    }
    
    public String getSTARTDATE(){
        return this.STARTDATE;
    }
    
    public String getENDDATE(){
        return this.ENDDATE;
    }
    
    public double getMONTHLYCOST(){
        return this.MONTHLYCOST;
    }
    
    public Tenant[] getRENTERS(){
        return this.RENTERS;
    }
    
    public House getRESIDENCE(){
        return this.RESIDENCE;
    }
    
    private String getBothParties(){
        
        StringBuilder allContractParties = new StringBuilder();
        
        allContractParties.append(RESIDENCE.getOwner());
        
        for(int i = 0; i < RENTERS.length; i++){
            allContractParties.append(" and " + RENTERS[i].getName());
        }
        return allContractParties.toString();
    }
}


// Dependency: RentalAgreementHardcopy relies on other objects to complete its behaviour. There's no printout without a rental agreement in the first place.
class RentalAgreementHardcopy{
        
    public void printAgreement(RentalAgreement agreement){
        // Not implemented...
        // Use file I/O and send to PDF...
    }
}

File: Lesson11_ClassRelationships/01_RelationshipsExample/RelationshipsExample.py
"""
Copyright Khawla Shnaikat and ChatGPT 3.5, 2023
Licensed under GPL v3
See LICENSE.txt for more information.
"""
class Roof:
    def __init__(self, material, installation_year):
        self.material = material
        self.installation_year = installation_year
        print(f"Roof built with {material} in {installation_year}.")

class Deck:
    def __init__(self, material, area, installation_year):
        self.material = material
        self.area = area
        self.installation_year = installation_year
        print("New deck built!")

class Door:
    def __init__(self, material="wood", colour="white"):
        self.material = material
        self.colour = colour
        print(f"Door created: Material - {material}, Colour - {colour}")

class Kitchen:
    def __init__(self, area, num_seats, num_appliances):
        self.area = area
        self.num_seats = num_seats
        self.num_appliances = num_appliances

class Bathroom:
    def __init__(self, area, sink_type, tub=None):
        self.area = area
        self.sink_type = sink_type
        self.tub = tub

    def add_tub(self, new_tub):
        self.tub = new_tub
        print("Bathroom tub added.")

    def set_sink(self, sink_type):
        self.sink_type = sink_type
        print("Bathroom sink set.")

class Bathtub:
    def __init__(self, type, colour):
        self.type = type
        self.colour = colour
        print(f"Bathtub added: Type - {type}, Colour - {colour}")

class Tenant:
    def __init__(self, name, occupation, phone_number, house_contact=None):
        self.name = name
        self.occupation = occupation
        self.phone_number = phone_number
        self.house_contact = house_contact

class RentalAgreement:
    def __init__(self, start_date, end_date, monthly_cost, renters, residence):
        self.start_date = start_date
        self.end_date = end_date
        self.monthly_cost = monthly_cost
        self.renters = renters
        self.residence = residence

    def rental_summary(self):
        print(f"Rental Agreement Summary: Start Date - {self.start_date}, End Date - {self.end_date}, Monthly Cost - {self.monthly_cost}")
        for renter in self.renters:
            print(f"Renter: {renter.name}, Occupation: {renter.occupation}, Phone: {renter.phone_number}")

class RentalAgreementHardcopy:
    def print_agreement(self, agreement):
        print("Printing rental agreement hardcopy.")
        
class RentalAgreementHardcopy:
    def print_agreement(self, agreement):
        print("Printing rental agreement hardcopy.")
        print(f"Rental Agreement: {agreement.start_date} to {agreement.end_date}")
        print(f"Monthly Cost: {agreement.monthly_cost}")
        print(f"Residence: {agreement.residence.address}")
        for renter in agreement.renters:
            print(f"Renter: {renter.name}, Occupation: {renter.occupation}, Contact: {renter.phone_number}")
       


class House:
    def __init__(self, owner, address, roof_material, roof_installed, kitchen_area, kitchen_seats, num_appliances, bathroom_area, sink_type, front_door=None, back_door=None):
        self.owner = owner
        self.address = address
        self.roof = Roof(roof_material, roof_installed)
        self.kitchen = Kitchen(kitchen_area, kitchen_seats, num_appliances)
        self.bathroom = Bathroom(bathroom_area, sink_type)
        self.front_door = front_door if front_door else Door()
        self.back_door = back_door if back_door else Door()
        self.deck = None
        print(f"House built at {address} by {owner}.")

    def update_roof(self, roof_material, roof_installed):
        self.roof = Roof(roof_material, roof_installed)
        print("Roof updated.")

    def renovate_bathroom(self, sink_type, new_tub):
        self.bathroom.add_tub(new_tub)
        self.bathroom.set_sink(sink_type)
        print("Bathroom renovated.")

    def build_deck(self, material, area, installation_year):
        self.deck = Deck(material, area, installation_year)
        print("Deck built.")

    def replace_door(self, which_door, new_door):
        if which_door.lower() == "front":
            self.front_door = new_door
        elif which_door.lower() == "back":
            self.back_door = new_door
        print(f"{which_door.capitalize()} door replaced.")

def main():
    new_house = House("Pat Neil", "123 Main St. NW", "cedar", "2010", 10.0, 6, 3, 4.5, "inset")
    new_door = Door("metal", "green")
    new_tub = Bathtub("clawfoot", "blue")
    new_house.build_deck("wood", 37.5, "2021")
    new_house.renovate_bathroom("farmhouse", new_tub)
    new_house.update_roof("asphalt", "2021")
    new_house.replace_door("front", new_door)
    renter1 = Tenant("Jordan Upton", "engineer", "587-289-1111")
    renter2 = Tenant("Jamie Upton", "teacher", "403-787-2222", renter1)
    contract = RentalAgreement("January 2021", "June 2021", 2500.00, [renter1, renter2], new_house)
    contract.rental_summary()
    legal_copy = RentalAgreementHardcopy()
    legal_copy.print_agreement(contract)

if __name__ == "__main__":
    main()


